### PR TITLE
ENH: Wrap ITKVideoStream

### PR DESCRIPTION
**Pull request by tbirdso**
Wraps `itk::VideoStream` and related components in `ITKVideoCore` for Python.

## PR Checklist
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5

### CODE REVIEW

File: Modules/Video/Core/wrapping/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+itk_wrap_module(ITKVideo)
+
+set(WRAPPER_SUBMODULE_ORDER
```
Comment (dzenanz) [inline Modules/Video/Core/wrapping/CMakeLists.txt:3]: This is only needed if order of compilation is important (something needs to go before something else).

File: Modules/Video/Core/wrapping/itkVideoStream.wrap [hunk 2]
```diff
@@ -0,0 +1,9 @@
+itk_wrap_include("itkImage.h")
+
+itk_wrap_class("itk::VideoStream" POINTER)
+  foreach(d ${ITK_WRAP_IMAGE_DIMS})
+    foreach(t ${WRAP_ITK_REAL})
```
[the remark this answers is not in the document]
Comment (tbirdso) [reply Modules/Video/Core/wrapping/itkVideoStream.wrap]: Good point. Fixed.

**Comment by tbirdso** [comment]
Thanks @dzenanz. I've added wrapping for `itk::TemporalDataObject`, which now fails with the warnings:
> 7>itkTemporalDataObject: warning(4): ITK type not wrapped, or currently not known: itk::DataObject
7>itkTemporalDataObject: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< itk::LightObject >
7>itkTemporalDataObject: warning(4): ITK type not wrapped, or currently not known: itk::TemporalDataObjectEnums::TemporalUnit
7>itkTemporalDataObject: warning(4): ITK type not wrapped, or currently not known: itk::TemporalRegion
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< itk::LightObject >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::RingBuffer< itk::DataObject >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Vector< double, 2 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Point< double, 2 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Matrix< double, 2, 2 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< double, 2 > >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::DataObject
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Vector< double, 3 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Point< double, 3 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::Matrix< double, 3, 3 >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< double, 3 > >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< float, 2 > >
7>itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< float, 3 > >

I can add wrappings for `itk::TemporalDataObjectEnums::TemporalUnit`, `itk::TemporalRegion`, and `itk::RingBuffer`, but I'm unclear on why the other warnings are present for classes wrapped in [ITKCommonBase.wrap](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/wrapping/ITKCommonBase.wrap). `itk-module.cmake` in ITKVideo explicitly sets ITKVideo to depend on ITKCommon:
> itk_module(ITKVideoCore
  ENABLE_SHARED
  DEPENDS
    ITKCommon

I've also verified in the Python console that `itk.DataObject` is available. Is there somewhere else we need to set wrapping orders so that ITKCommon is built first?

File: Modules/Video/Core/wrapping/CMakeLists.txt (hunk 1 above)
Comment (tbirdso) [reply Modules/Video/Core/wrapping/CMakeLists.txt:3]: Understood - order of compilation is important now that we're adding parent class wrappings as well.

File: Modules/Video/Core/wrapping/itkVideoStream.wrap [hunk 3]
```diff
@@ -0,0 +1,9 @@
+itk_wrap_include("itkImage.h")
```
Comment (dzenanz) [inline Modules/Video/Core/wrapping/itkVideoStream.wrap:1]: Maybe itk_wrap_include("itkDataObject") is needed?

**Comment by thewtex** [comment]
@tbirdso yes, we are not importing the swig definitions from ITKCommon, which is causing these errors.

Digging into this, the issues is the `itk_wrap_module(ITKVideo)` in *Modules/Video/Core/wrapping/CMakeLists.txt*. This should be `itk_wrap_module(ITKVideoCore)` following the module name defined in *Modules/Video/Core/itk-module.cmake*.

**Comment by tbirdso** [comment]
Thank you @thewtex! Fixing the name mismatch solved the issue. Squashed commits.

**Review by dzenanz — APPROVED** [review]

**Comment by tbirdso** [comment]
Adding wrapping for `itk::TemporalRegion` to address build warnings at [https://open.cdash.org/viewBuildError.php?type=1&buildid=7232331](https://open.cdash.org/viewBuildError.php?type=1&buildid=7232331).

**Review by thewtex — CHANGES_REQUESTED** [review]
Excellent! 🐎 

We should also wrap TemporalDataObjectEnums while wrapping TemporalDataObject. `git grep Enums | grep wrap` for examples.

**Comment by thewtex** [comment]
What are the remaining CI issues here?

**Comment by tbirdso** [comment]
Sorry, this one fell off my radar. Build is succeeding on Windows, the `ITK.Linux.Python` and `ITK.macOS.Python` CI don't seem to have any failure information. I will rebuild locally on Linux and Windows and re-run CI to get failure text.

**Comment by tbirdso** [comment]
CI has several warnings regarding wrapping with `itk::SmartPointer< const itk::Image<>>` types:
> itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< float, 2 > >
> itkVideoStream: warning(4): ITK type not wrapped, or currently not known: itk::SmartPointer< const itk::Image< float, 3 > >
...

**Comment by thewtex** [comment]
If the return arguments here:

https://github.com/InsightSoftwareConsortium/ITK/blob/71bb6a99fe02fe80e8e39ed4e00b009b3c7444dd/Modules/Video/Core/include/itkVideoStream.h#L190-L193

are changed from `SmartPointer`'s to raw pointers, is the issue addressed?

**Comment by thewtex** [comment]
WIP -> BUG

**Comment by tbirdso** [comment]
Squashed branch history following green CI for fixes.

**Comment by tbirdso** [comment]
Observed new macOS failure even though no new changes were made in previous push:
> The following tests FAILED:
	231 - itkPyBufferMemoryLeakTest (Failed)

Re-running CI to see if we can reproduce the issue.

**Comment by thewtex** [comment]
@tbirdso could this please be rebased on `master`?

**Review by thewtex — APPROVED** [review]

**Pull request merged**